File: skills/sci-plot/assets/case-packs/rf-0118/plot_sankey.py
from pathlib import Path
from collections import OrderedDict
import colorsys
import numpy as np
from openpyxl import load_workbook
import matplotlib.pyplot as plt
from matplotlib.patches import Polygon, Rectangle
from matplotlib.colors import to_rgb


def read_xlsx(path):
    workbook = load_workbook(path, read_only=True, data_only=True)
    worksheet = workbook.active
    iterator = worksheet.iter_rows(values_only=True)
    headers = [str(value) if value is not None else "" for value in next(iterator)]
    records = [dict(zip(headers, row)) for row in iterator if any(value is not None for value in row)]
    workbook.close()
    return records, headers


def require_columns(headers, columns, source):
    missing = [name for name in columns if name not in headers]
    if missing:
        raise ValueError(f"{source} is missing required columns: {', '.join(missing)}")


def unique(values):
    return list(OrderedDict.fromkeys(values))


def totals_by(records, column, weight="_weight"):
    totals = OrderedDict()
    for row in records:
        key = row[column]
        totals[key] = totals.get(key, 0.0) + float(row[weight])
    return totals


def palette(size, start):
    return [colorsys.hls_to_rgb((start + i * 0.61803398875) % 1.0, 0.56, 0.52) for i in range(size)]


def adjust_positions(values, minimum, lower=0.025, upper=0.975):
    values = np.asarray(values, dtype=float)
    order = np.argsort(values)
    placed = np.clip(values[order], lower, upper)
    for i in range(1, len(placed)):
        placed[i] = max(placed[i], placed[i - 1] + minimum)
    if len(placed) and placed[-1] > upper:
        placed -= placed[-1] - upper
        for i in range(len(placed) - 2, -1, -1):
            placed[i] = min(placed[i], placed[i + 1] - minimum)
    if len(placed) and placed[0] < lower:
        placed += lower - placed[0]
    result = np.empty_like(placed)
    result[order] = placed
    return result


def build_flow(records, stages, orders, weight_column=None):
    data = []
    for record in records:
        row = dict(record)
        weight = float(row[weight_column]) if weight_column else 1.0
        if not np.isfinite(weight) or weight <= 0:
            raise ValueError("All flow weights must be finite and positive")
        row["_weight"] = weight
        data.append(row)
    total = sum(row["_weight"] for row in data)
    stage_orders = []
    for column, requested in zip(stages, orders):
        observed = unique(str(row[column]) for row in data)
        ordered = [value for value in requested if value in observed]
        ordered.extend(value for value in observed if value not in ordered)
        stage_orders.append(ordered)
    maximum_nodes = max(map(len, stage_orders))
    gap = 0.006 if maximum_nodes > 24 else 0.012
    scale = (0.88 - gap * (maximum_nodes - 1)) / total
    nodes = []
    starts = [0.57, 0.03, 0.42, 0.77]
    for stage_index, (column, ordered) in enumerate(zip(stages, stage_orders)):
        totals = totals_by(data, column)
        colors = palette(len(ordered), starts[stage_index % len(starts)])
        occupied = total * scale + gap * (len(ordered) - 1)
        cursor = 0.5 + occupied / 2
        stage_nodes = []
        for index, name in enumerate(ordered):
            value = float(totals[name])
            stage_nodes.append({"name": name, "value": value, "y1": cursor, "y0": cursor - value * scale, "color": colors[index]})
            cursor -= value * scale + gap
        nodes.append(stage_nodes)
    links_by_gap = []
    for stage_index in range(len(stages) - 1):
        source_column = stages[stage_index]
        target_column = stages[stage_index + 1]
        grouped = OrderedDict()
        for row in data:
            key = (row[source_column], row[target_column])
            grouped[key] = grouped.get(key, 0.0) + row["_weight"]
        source_index = {node["name"]: i for i, node in enumerate(nodes[stage_index])}
        target_index = {node["name"]: i for i, node in enumerate(nodes[stage_index + 1])}
        links = [{"source": source, "target": target, "weight": weight, "source_index": source_index[source], "target_index": target_index[target]} for (source, target), weight in grouped.items()]
        source_cursor = {node["name"]: node["y1"] for node in nodes[stage_index]}
        for link in sorted(links, key=lambda item: (item["source_index"], item["target_index"])):
            height = link["weight"] * scale
            link["sy1"] = source_cursor[link["source"]]
            link["sy0"] = source_cursor[link["source"]] - height
            source_cursor[link["source"]] -= height
        target_cursor = {node["name"]: node["y1"] for node in nodes[stage_index + 1]}
        for link in sorted(links, key=lambda item: (item["target_index"], item["source_index"])):
            height = link["weight"] * scale
            link["ty1"] = target_cursor[link["target"]]
            link["ty0"] = target_cursor[link["target"]] - height
            target_cursor[link["target"]] -= height
        links_by_gap.append(links)
    return {"stages": stages, "nodes": nodes, "links": links_by_gap, "total": total}


def draw_ribbon(ax, x0, x1, link, source_color, target_color, segments=30):
    t = np.linspace(0, 1, segments + 1)
    smooth = t * t * (3 - 2 * t)
    xs = x0 + (x1 - x0) * t
    upper = link["sy1"] + (link["ty1"] - link["sy1"]) * smooth
    lower = link["sy0"] + (link["ty0"] - link["sy0"]) * smooth
    source = np.asarray(to_rgb(source_color))
    target = np.asarray(to_rgb(target_color))
    for i in range(segments):
        fraction = (i + 0.5) / segments
        color = source * (1 - fraction) + target * fraction
        polygon = Polygon([[xs[i], upper[i]], [xs[i + 1], upper[i + 1]], [xs[i + 1], lower[i + 1]], [xs[i], lower[i]]], closed=True, facecolor=color, edgecolor="none", alpha=0.52, rasterized=True)
        ax.add_patch(polygon)


def format_value(value, integer):
    return f"{int(round(value)):,}" if integer else f"{value:.1f}"


def draw_flow(ax, flow, headers, title, subtitle, integer):
    stage_count = len(flow["stages"])
    x_positions = np.arange(stage_count, dtype=float)
    node_width = 0.055
    node_lookup = [{node["name"]: node for node in stage} for stage in flow["nodes"]]
    for stage_index, links in enumerate(flow["links"]):
        for link in sorted(links, key=lambda item: item["weight"], reverse=True):
            draw_ribbon(ax, x_positions[stage_index] + node_width / 2, x_positions[stage_index + 1] - node_width / 2, link, node_lookup[stage_index][link["source"]]["color"], node_lookup[stage_index + 1][link["target"]]["color"])
    for stage_index, stage_nodes in enumerate(flow["nodes"]):
        centers = [(node["y0"] + node["y1"]) / 2 for node in stage_nodes]
        minimum = 0.026 if len(stage_nodes) > 20 else 0.032
        labels = adjust_positions(centers, minimum)
        for node, label_y, center_y in zip(stage_nodes, labels, centers):
            height = max(node["y1"] - node["y0"], 0.0011)
            center = (node["y0"] + node["y1"]) / 2
            ax.add_patch(Rectangle((x_positions[stage_index] - node_width / 2, center - height / 2), node_width, height, facecolor=node["color"], edgecolor="#FFFFFF", linewidth=0.45, zorder=4))
            label = f"{node['name']}  {format_value(node['value'], integer)}"
            if stage_index == 0:
                text_x = x_positions[stage_index] - node_width / 2 - 0.025
                horizontal = "right"
                anchor_x = x_positions[stage_index] - node_width / 2
            else:
                text_x = x_positions[stage_index] + node_width / 2 + 0.025
                horizontal = "left"
                anchor_x = x_positions[stage_index] + node_width / 2
            if abs(label_y - center_y) > 0.004:
                ax.plot([anchor_x, text_x], [center_y, label_y], color="#8B959D", linewidth=0.38, zorder=5)
            ax.text(text_x, label_y, label, ha=horizontal, va="center", fontsize=6.2 if len(stage_nodes) > 20 else 7.2, color="#17222B", zorder=6, bbox={"boxstyle": "round,pad=0.12", "facecolor": "#FFFFFF", "edgecolor": "none", "alpha": 0.78})
        ax.text(x_positions[stage_index], 1.016, headers[stage_index].upper(), ha="center", va="bottom", fontsize=8.1, weight="bold", color="#52616B")
    ax.text(0, 1.105, title, transform=ax.transAxes, ha="left", va="bottom", fontsize=14.5, weight="bold", color="#15252D")
    ax.text(0, 1.058, subtitle, transform=ax.transAxes, ha="left", va="bottom", fontsize=8.7, color="#53636C")
    ax.set_xlim(-0.48, stage_count - 1 + 0.82)
    ax.set_ylim(-0.03, 1.055)
    ax.axis("off")


base = Path(__file__).resolve().parent
data1, headers1 = read_xlsx(base / "data1.xlsx")
data2, headers2 = read_xlsx(base / "data2.xlsx")
require_columns(headers1, ["Timeline", "Lineages Distribution", "Geographic Location", "P.M.A.s"], "data1.xlsx")
require_columns(headers2, ["cluster", "mRNA1", "mRNA2", "Freq"], "data2.xlsx")
if any(row["Timeline"] is None or row["Lineages Distribution"] is None or row["Geographic Location"] is None for row in data1):
    raise ValueError("data1.xlsx contains missing values in the first three stages")
for row in data1:
    row["Timeline"] = str(row["Timeline"])
    row["Lineages Distribution"] = str(row["Lineages Distribution"])
    row["Geographic Location"] = str(row["Geographic Location"])
    row["P.M.A.s / status"] = "Outside China / not applicable" if row["P.M.A.s"] is None else str(row["P.M.A.s"])
for row in data2:
    if any(row[column] is None for column in ["cluster", "mRNA1", "mRNA2", "Freq"]):
        raise ValueError("data2.xlsx contains missing flow values")
    for column in ["cluster", "mRNA1", "mRNA2"]:
        row[column] = str(row[column])
timeline_order = sorted(unique(row["Timeline"] for row in data1), key=lambda value: int(value.split("-")[0]))
lineage_order = [value for value in ["Paraphyletic cluster", "Sublineage 8.7"] if value in set(row["Lineages Distribution"] for row in data1)]
geographic_totals = totals_by([dict(row, _weight=1.0) for row in data1], "Geographic Location")
geographic_order = [value for value, _ in sorted(geographic_totals.items(), key=lambda item: item[1], reverse=True) if value != "China"] + (["China"] if "China" in geographic_totals else [])
pma_totals = totals_by([dict(row, _weight=1.0) for row in data1], "P.M.A.s / status")
pma_order = (["Outside China / not applicable"] if "Outside China / not applicable" in pma_totals else []) + [value for value, _ in sorted(pma_totals.items(), key=lambda item: item[1], reverse=True) if value != "Outside China / not applicable"]
flow1 = build_flow(data1, ["Timeline", "Lineages Distribution", "Geographic Location", "P.M.A.s / status"], [timeline_order, lineage_order, geographic_order, pma_order])
weighted2 = [dict(row, _weight=float(row["Freq"])) for row in data2]
cluster_order = [value for value, _ in sorted(totals_by(weighted2, "cluster").items(), key=lambda item: item[1], reverse=True)]
mrna1_order = [value for value, _ in sorted(totals_by(weighted2, "mRNA1").items(), key=lambda item: item[1], reverse=True)]
mrna2_order = [value for value, _ in sorted(totals_by(weighted2, "mRNA2").items(), key=lambda item: item[1], reverse=True)]
flow2 = build_flow(data2, ["cluster", "mRNA1", "mRNA2"], [cluster_order, mrna1_order, mrna2_order], "Freq")
fig, axes = plt.subplots(1, 2, figsize=(19, 11.5), gridspec_kw={"width_ratios": [1.16, 0.84]})
fig.patch.set_facecolor("#F8F7F3")
for axis in axes:
    axis.set_facecolor("#F8F7F3")
draw_flow(axes[0], flow1, ["Time window", "Lineage", "Country", "P.M.A.s / status"], "A  Isolate distribution across time, lineage and geography", f"data1.xlsx · n = {len(data1):,} records · ribbon width = record count", True)
sum_freq = sum(float(row["Freq"]) for row in data2)
draw_flow(axes[1], flow2, ["Cluster", "mRNA 1", "mRNA 2"], "B  Weighted cluster–mRNA associations", f"data2.xlsx · {len(data2):,} rows · Σ Freq = {sum_freq:.2f} · ribbon width = Freq", False)
missing_pma = sum(row["P.M.A.s"] is None for row in data1)
fig.suptitle("Two independent alluvial datasets", x=0.055, y=0.985, ha="left", fontsize=20, weight="bold", color="#10242C")
fig.text(0.055, 0.953, "Panels are intentionally separate: the workbooks contain no shared record-level key and use different flow measures.", ha="left", va="top", fontsize=10.2, color="#53636C")
fig.text(0.055, 0.018, f"Blank P.M.A.s values are retained as “Outside China / not applicable” (n = {missing_pma:,}; every blank occurs outside China). Node values are totals within each stage.", ha="left", va="bottom", fontsize=8.5, color="#5D6970")
fig.subplots_adjust(left=0.045, right=0.987, top=0.79, bottom=0.055, wspace=0.10)
output = base / "case29_sankey_python.png"
fig.savefig(output, dpi=360, facecolor=fig.get_facecolor(), bbox_inches="tight", pad_inches=0.12)
plt.close(fig)
print(output)

File: skills/sci-plot/assets/case-packs/rf-0118/plot_sankey.R
script_path <- local({
  frames <- sys.frames()
  paths <- vapply(frames, function(frame) if (is.null(frame$ofile)) "" else as.character(frame$ofile), character(1))
  paths <- paths[nzchar(paths)]
  if (length(paths)) return(normalizePath(tail(paths, 1), winslash = "/", mustWork = TRUE))
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- sub("^--file=", "", args[grepl("^--file=", args)])
  if (length(file_arg)) return(normalizePath(file_arg[[1]], winslash = "/", mustWork = TRUE))
  stop("Cannot determine the script path")
})
base_dir <- dirname(script_path)
if (!requireNamespace("readxl", quietly = TRUE)) stop("Package 'readxl' is required")
require_columns <- function(data, columns, source) {
  missing <- setdiff(columns, names(data))
  if (length(missing)) stop(sprintf("%s is missing required columns: %s", source, paste(missing, collapse = ", ")))
}
ordered_unique <- function(values) values[!duplicated(values)]
totals_by <- function(data, column) {
  result <- tapply(data$.weight, data[[column]], sum)
  result[order(result, decreasing = TRUE)]
}
make_palette <- function(size, start) {
  if (!size) return(character())
  grDevices::hcl(h = (start + (seq_len(size) - 1) * 222.492) %% 360, c = 55, l = 60)
}
adjust_positions <- function(values, minimum, lower = 0.025, upper = 0.975) {
  order_index <- order(values)
  placed <- pmin(pmax(values[order_index], lower), upper)
  if (length(placed) > 1) {
    for (i in 2:length(placed)) placed[i] <- max(placed[i], placed[i - 1] + minimum)
  }
  if (length(placed) && tail(placed, 1) > upper) {
    placed <- placed - (tail(placed, 1) - upper)
    if (length(placed) > 1) {
      for (i in seq(length(placed) - 1, 1)) placed[i] <- min(placed[i], placed[i + 1] - minimum)
    }
  }
  if (length(placed) && placed[1] < lower) placed <- placed + lower - placed[1]
  result <- numeric(length(values))
  result[order_index] <- placed
  result
}
build_flow <- function(data, stages, orders, weight_column = NULL) {
  data <- as.data.frame(data, stringsAsFactors = FALSE)
  for (column in stages) data[[column]] <- as.character(data[[column]])
  data$.weight <- if (is.null(weight_column)) rep(1, nrow(data)) else as.numeric(data[[weight_column]])
  if (any(!is.finite(data$.weight)) || any(data$.weight <= 0)) stop("All flow weights must be finite and positive")
  total <- sum(data$.weight)
  stage_orders <- vector("list", length(stages))
  for (i in seq_along(stages)) {
    observed <- ordered_unique(data[[stages[i]]])
    stage_orders[[i]] <- c(orders[[i]][orders[[i]] %in% observed], observed[!observed %in% orders[[i]]])
  }
  maximum_nodes <- max(lengths(stage_orders))
  gap <- if (maximum_nodes > 24) 0.006 else 0.012
  scale <- (0.88 - gap * (maximum_nodes - 1)) / total
  starts <- c(205, 11, 151, 277)
  nodes <- vector("list", length(stages))
  for (i in seq_along(stages)) {
    totals <- tapply(data$.weight, data[[stages[i]]], sum)
    order_values <- stage_orders[[i]]
    values <- as.numeric(totals[order_values])
    colors <- make_palette(length(order_values), starts[(i - 1) %% length(starts) + 1])
    occupied <- total * scale + gap * (length(order_values) - 1)
    cursor <- 0.5 + occupied / 2
    y1 <- numeric(length(order_values))
    y0 <- numeric(length(order_values))
    for (j in seq_along(order_values)) {
      y1[j] <- cursor
      y0[j] <- cursor - values[j] * scale
      cursor <- y0[j] - gap
    }
    nodes[[i]] <- data.frame(name = order_values, value = values, y1 = y1, y0 = y0, color = colors, stringsAsFactors = FALSE)
  }
  links_by_gap <- vector("list", length(stages) - 1)
  for (i in seq_len(length(stages) - 1)) {
    source_column <- stages[i]
    target_column <- stages[i + 1]
    links <- aggregate(data$.weight, by = list(source = data[[source_column]], target = data[[target_column]]), FUN = sum)
    names(links)[3] <- "weight"
    links$source_index <- match(links$source, nodes[[i]]$name)
    links$target_index <- match(links$target, nodes[[i + 1]]$name)
    links$sy1 <- links$sy0 <- links$ty1 <- links$ty0 <- NA_real_
    source_cursor <- stats::setNames(nodes[[i]]$y1, nodes[[i]]$name)
    for (row_index in order(links$source_index, links$target_index)) {
      source <- links$source[row_index]
      height <- links$weight[row_index] * scale
      links$sy1[row_index] <- source_cursor[[source]]
      links$sy0[row_index] <- source_cursor[[source]] - height
      source_cursor[[source]] <- source_cursor[[source]] - height
    }
    target_cursor <- stats::setNames(nodes[[i + 1]]$y1, nodes[[i + 1]]$name)
    for (row_index in order(links$target_index, links$source_index)) {
      target <- links$target[row_index]
      height <- links$weight[row_index] * scale
      links$ty1[row_index] <- target_cursor[[target]]
      links$ty0[row_index] <- target_cursor[[target]] - height
      target_cursor[[target]] <- target_cursor[[target]] - height
    }
    links_by_gap[[i]] <- links
  }
  list(stages = stages, nodes = nodes, links = links_by_gap, total = total)
}
draw_ribbon <- function(x0, x1, row, source_color, target_color, segments = 30) {
  t <- seq(0, 1, length.out = segments + 1)
  smooth <- t * t * (3 - 2 * t)
  x <- x0 + (x1 - x0) * t
  upper <- row$sy1 + (row$ty1 - row$sy1) * smooth
  lower <- row$sy0 + (row$ty0 - row$sy0) * smooth
  colors <- grDevices::colorRampPalette(c(source_color, target_color), space = "Lab")(segments)
  colors <- grDevices::adjustcolor(colors, alpha.f = 0.52)
  for (i in seq_len(segments)) {
    polygon(c(x[i], x[i + 1], x[i + 1], x[i]), c(upper[i], upper[i + 1], lower[i + 1], lower[i]), col = colors[i], border = NA)
  }
}
format_value <- function(value, integer) {
  if (integer) format(round(value), big.mark = ",", scientific = FALSE, trim = TRUE) else sprintf("%.1f", value)
}
draw_label <- function(x, y, label, align, cex) {
  width <- strwidth(label, cex = cex)
  height <- strheight(label, cex = cex)
  if (align == 0) {
    rect(x - 0.006, y - height * 0.62, x + width + 0.012, y + height * 0.62, col = grDevices::adjustcolor("#FFFFFF", alpha.f = 0.79), border = NA)
  } else {
    rect(x - width - 0.012, y - height * 0.62, x + 0.006, y + height * 0.62, col = grDevices::adjustcolor("#FFFFFF", alpha.f = 0.79), border = NA)
  }
  text(x, y, label, adj = c(align, 0.5), cex = cex, col = "#17222B")
}
draw_flow <- function(flow, headers, panel_title, subtitle, integer) {
  stage_count <- length(flow$stages)
  x_positions <- seq(0, stage_count - 1)
  node_width <- 0.055
  plot.new()
  plot.window(xlim = c(-0.48, stage_count - 1 + 0.82), ylim = c(-0.03, 1.055), xaxs = "i", yaxs = "i")
  rect(par("usr")[1], par("usr")[3], par("usr")[2], par("usr")[4], col = "#F8F7F3", border = NA)
  for (stage_index in seq_along(flow$links)) {
    links <- flow$links[[stage_index]]
    source_nodes <- flow$nodes[[stage_index]]
    target_nodes <- flow$nodes[[stage_index + 1]]
    for (row_index in order(links$weight, decreasing = TRUE)) {
      row <- links[row_index, , drop = FALSE]
      source_color <- source_nodes$color[match(row$source, source_nodes$name)]
      target_color <- target_nodes$color[match(row$target, target_nodes$name)]
      draw_ribbon(x_positions[stage_index] + node_width / 2, x_positions[stage_index + 1] - node_width / 2, row, source_color, target_color)
    }
  }
  for (stage_index in seq_along(flow$nodes)) {
    stage_nodes <- flow$nodes[[stage_index]]
    centers <- (stage_nodes$y0 + stage_nodes$y1) / 2
    minimum <- if (nrow(stage_nodes) > 20) 0.026 else 0.032
    label_positions <- adjust_positions(centers, minimum)
    for (i in seq_len(nrow(stage_nodes))) {
      height <- max(stage_nodes$y1[i] - stage_nodes$y0[i], 0.0011)
      center <- centers[i]
      rect(x_positions[stage_index] - node_width / 2, center - height / 2, x_positions[stage_index] + node_width / 2, center + height / 2, col = stage_nodes$color[i], border = "#FFFFFF", lwd = 0.45)
      label <- paste0(stage_nodes$name[i], "  ", format_value(stage_nodes$value[i], integer))
      if (stage_index == 1) {
        text_x <- x_positions[stage_index] - node_width / 2 - 0.025
        align <- 1
        anchor_x <- x_positions[stage_index] - node_width / 2
      } else {
        text_x <- x_positions[stage_index] + node_width / 2 + 0.025
        align <- 0
        anchor_x <- x_positions[stage_index] + node_width / 2
      }
      if (abs(label_positions[i] - center) > 0.004) segments(anchor_x, center, text_x, label_positions[i], col = "#8B959D", lwd = 0.38)
      draw_label(text_x, label_positions[i], label, align, if (nrow(stage_nodes) > 20) 0.48 else 0.58)
    }
    text(x_positions[stage_index], 1.016, toupper(headers[stage_index]), adj = c(0.5, 0), cex = 0.68, font = 2, col = "#52616B")
  }
  mtext(panel_title, side = 3, line = 2.5, adj = 0, cex = 1.04, font = 2, col = "#15252D")
  mtext(subtitle, side = 3, line = 1.25, adj = 0, cex = 0.69, col = "#53636C")
}
data1 <- as.data.frame(readxl::read_xlsx(file.path(base_dir, "data1.xlsx")), stringsAsFactors = FALSE)
data2 <- as.data.frame(readxl::read_xlsx(file.path(base_dir, "data2.xlsx")), stringsAsFactors = FALSE)
require_columns(data1, c("Timeline", "Lineages Distribution", "Geographic Location", "P.M.A.s"), "data1.xlsx")
require_columns(data2, c("cluster", "mRNA1", "mRNA2", "Freq"), "data2.xlsx")
if (anyNA(data1[c("Timeline", "Lineages Distribution", "Geographic Location")])) stop("data1.xlsx contains missing values in the first three stages")
data1$Timeline <- as.character(data1$Timeline)
data1$`Lineages Distribution` <- as.character(data1$`Lineages Distribution`)
data1$`Geographic Location` <- as.character(data1$`Geographic Location`)
data1$`P.M.A.s / status` <- ifelse(is.na(data1$P.M.A.s), "Outside China / not applicable", as.character(data1$P.M.A.s))
if (anyNA(data2[c("cluster", "mRNA1", "mRNA2", "Freq")])) stop("data2.xlsx contains missing flow values")
data2$cluster <- as.character(data2$cluster)
data2$mRNA1 <- as.character(data2$mRNA1)
data2$mRNA2 <- as.character(data2$mRNA2)
data1$.weight <- 1
timeline_order <- sort(unique(data1$Timeline))
lineage_order <- intersect(c("Paraphyletic cluster", "Sublineage 8.7"), unique(data1$`Lineages Distribution`))
geographic_totals <- totals_by(data1, "Geographic Location")
geographic_order <- c(setdiff(names(geographic_totals), "China"), intersect("China", names(geographic_totals)))
pma_totals <- totals_by(data1, "P.M.A.s / status")
pma_order <- c(intersect("Outside China / not applicable", names(pma_totals)), setdiff(names(pma_totals), "Outside China / not applicable"))
flow1 <- build_flow(data1, c("Timeline", "Lineages Distribution", "Geographic Location", "P.M.A.s / status"), list(timeline_order, lineage_order, geographic_order, pma_order))
data2$.weight <- as.numeric(data2$Freq)
cluster_order <- names(totals_by(data2, "cluster"))
mrna1_order <- names(totals_by(data2, "mRNA1"))
mrna2_order <- names(totals_by(data2, "mRNA2"))
flow2 <- build_flow(data2, c("cluster", "mRNA1", "mRNA2"), list(cluster_order, mrna1_order, mrna2_order), "Freq")
output <- file.path(base_dir, "case29_sankey_r.png")
if (requireNamespace("ragg", quietly = TRUE)) {
  ragg::agg_png(output, width = 19, height = 11.5, units = "in", res = 360, background = "#F8F7F3")
} else if (capabilities("cairo")) {
  grDevices::png(output, width = 19, height = 11.5, units = "in", res = 360, bg = "#F8F7F3", type = "cairo-png")
} else {
  grDevices::png(output, width = 19, height = 11.5, units = "in", res = 360, bg = "#F8F7F3")
}
layout(matrix(1:2, nrow = 1), widths = c(1.16, 0.84))
par(oma = c(2.6, 1.1, 5.2, 1.1), mar = c(0.2, 0.2, 4.1, 0.2), family = "sans")
draw_flow(flow1, c("Time window", "Lineage", "Country", "P.M.A.s / status"), "A  Isolate distribution across time, lineage and geography", sprintf("data1.xlsx · n = %s records · ribbon width = record count", format(nrow(data1), big.mark = ",")), TRUE)
draw_flow(flow2, c("Cluster", "mRNA 1", "mRNA 2"), "B  Weighted cluster–mRNA associations", sprintf("data2.xlsx · %s rows · Σ Freq = %.2f · ribbon width = Freq", nrow(data2), sum(data2$Freq)), FALSE)
mtext("Two independent alluvial datasets", side = 3, outer = TRUE, line = 3.45, adj = 0.015, cex = 1.52, font = 2, col = "#10242C")
mtext("Panels are intentionally separate: the workbooks contain no shared record-level key and use different flow measures.", side = 3, outer = TRUE, line = 1.85, adj = 0.015, cex = 0.82, col = "#53636C")
mtext(sprintf("Blank P.M.A.s values are retained as “Outside China / not applicable” (n = %s; every blank occurs outside China). Node values are totals within each stage.", format(sum(is.na(data1$P.M.A.s)), big.mark = ",")), side = 1, outer = TRUE, line = 0.75, adj = 0.015, cex = 0.68, col = "#5D6970")
invisible(dev.off())
cat(output, "\n")
